Add hotsos documentation (#652)
canonical/hotsos: doc/source/contrib/testing.rst

Testing Checks
==============

All checks contributed to hotsos should be accompanied by a unit test and there are
two ways to do this. You can write tests the "old" way i.e. as a Python unittest
implementation under *tests/unit_tests* or you can write them in yaml
which is explained here.

Writing tests in Yaml
---------------------

Checks written in yaml can have one or more associated unit test also written
in yaml. These will be loaded at runtime into the main unittest runner and run
as part of the existing Python unit tests.

To write a test you must mimic the path to the check under test i.e. the "target".

> For example lets say we have a written a scenario check add a single test as follows::

Reviewer: ```suggestion
For example given a scenario check we can add a single test as follows:
```

    hotsos/defs/scenarios/myplugin/foo.yaml
    hotsos/defs/tests/scenarios/myplugin/foo.yaml

The point of using the same sub path and filename is that the runner will then assume that the
target has the same name.

To write more than one test for a given target, perhaps to test both pass and fail cases, you
can give the test file a meaningful name and inside your test specific the target file name

> using the ``target-name: <name>.yaml`` stanza e.g.::

Reviewer: ```suggestion
using the ``target-name: <name>.yaml`` stanza e.g.:
```
